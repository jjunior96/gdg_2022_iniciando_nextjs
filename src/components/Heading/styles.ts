import styled, { css, DefaultTheme } from 'styled-components';

import { HeadingProps, LineColorProps } from '.';

const containerModifiers = {
  lineLeft: (theme: DefaultTheme, lineColor: LineColorProps) => css`
    padding-left: ${theme.spacings.xxsmall};

    border-left: 0.7rem solid ${theme.colors[lineColor]};
  `,

  lineBottom: (theme: DefaultTheme, lineColor: LineColorProps) => css`
    position: relative;
    margin-bottom: ${theme.spacings.medium};

    &::after {
      content: '';
      position: absolute;
      left: 0;
      bottom: -0.5rem;
      width: 5rem;
      border-bottom: 0.5rem solid ${theme.colors[lineColor]};
    }
  `,

  small: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.xsmall};
  `,

  medium: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.small};
  `,

  large: (theme: DefaultTheme) => css`
    font-size: ${theme.font.sizes.large};
  `
};

export const Container = styled.h2<HeadingProps>`
  ${({ theme, color, lineLeft, lineBottom, size, lineColor }) => css`
    color: ${theme.colors[color!]};

    ${!!size && containerModifiers[size](theme)};

    ${lineLeft && containerModifiers.lineLeft(theme, lineColor!)}
    ${lineBottom && containerModifiers.lineBottom(theme, lineColor!)}
  `}
`;
