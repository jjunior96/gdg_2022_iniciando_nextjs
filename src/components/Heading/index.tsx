import * as S from './styles';

export type LineColorProps = 'primary' | 'secundary';

export type HeadingProps = {
  children: React.ReactNode;
  color?: 'white' | 'black';
  lineLeft?: boolean;
  lineBottom?: boolean;
  size?: 'small' | 'medium' | 'large';
  lineColor?: LineColorProps;
};

const Heading = ({
  children,
  color,
  lineLeft = false,
  lineBottom = false,
  size = 'large',
  lineColor = 'primary'
}: HeadingProps) => {
  return (
    <S.Container
      color={color}
      lineLeft={lineLeft}
      size={size}
      lineColor={lineColor}
      lineBottom={lineBottom}
    >
      {children}
    </S.Container>
  );
};

export default Heading;
