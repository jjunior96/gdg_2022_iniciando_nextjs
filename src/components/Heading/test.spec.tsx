import { screen } from '@testing-library/react';

import { renderWithTheme } from 'utils/tests/__config__/helpers';

import Heading from '.';

describe('<Heading />', () => {
  it('should render the heading', () => {
    const { container } = renderWithTheme(<Heading>Text Example</Heading>);

    expect(
      screen.getByRole('heading', { name: /Text Example/i })
    ).toBeInTheDocument();
    expect(screen.getByRole('heading', { name: /Text Example/i })).toHaveStyle({
      'font-size': '3.2rem'
    });

    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the heading with small size', () => {
    renderWithTheme(<Heading size="small">Text Example</Heading>);

    expect(screen.getByRole('heading', { name: /Text Example/i })).toHaveStyle({
      'font-size': '1.6rem'
    });
  });

  it('should render the heading with medium size', () => {
    renderWithTheme(<Heading size="medium">Text Example</Heading>);

    expect(screen.getByRole('heading', { name: /Text Example/i })).toHaveStyle({
      'font-size': '2rem'
    });
  });

  it('should render the heading with large size', () => {
    renderWithTheme(<Heading size="large">Text Example</Heading>);

    expect(screen.getByRole('heading', { name: /Text Example/i })).toHaveStyle({
      'font-size': '3.2rem'
    });
  });

  it('should render the heading with lineBottom prop', () => {
    renderWithTheme(
      <Heading size="large" lineBottom>
        Text Example
      </Heading>
    );

    expect(screen.getByRole('heading', { name: /Text Example/i })).toHaveStyle({
      position: 'relative',
      'margin-bottom': '3.2rem'
    });
  });
});
