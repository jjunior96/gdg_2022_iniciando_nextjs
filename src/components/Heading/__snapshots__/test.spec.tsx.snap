// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Heading /> should render the heading 1`] = `
.c0 {
  font-size: 3.2rem;
}

<h2
  class="c0"
>
  Text Example
</h2>
`;
